x/leverage/keeper/oracle.go

```diff
@@ -12,12 +12,33 @@ import (
 // the base and display/symbol denominations for each exchange pair. E.g. it must
 // know about the UMEE/USD exchange rate along with the uumee base denomination
 // and the exponent.
-func (k Keeper) TokenPrice(ctx sdk.Context, denom string) (sdk.Dec, error) {
+// This function will not panic or return non-positive exchange rates, preferring
+// to error instead.
+func (k Keeper) TokenPrice(ctx sdk.Context, denom string) (price sdk.Dec, err error) {
+
 	if !k.IsAcceptedToken(ctx, denom) {
 		return sdk.ZeroDec(), sdkerrors.Wrap(types.ErrInvalidAsset, denom)
 	}
 
-	return k.oracleKeeper.GetExchangeRateBase(ctx, denom)
+	defer func() {
+		// This will catch panics from the function body
+		// and instead return a general error.
+		if pan := recover(); pan != nil {
```

Comment:
This is not good -- where does the panic come from? I would much rather prefer to not panic in these code path(s) and instead return an error.

Reply:
The panic would be coming from the lines in the function body - so the call to the oracle module.

Changing the caught panic into an error here allows us to distinguish between oracle-related and leverage-related panics, as the current state of things gives us something like "panic message hidden to protect potentially sensitive content".

Hopefully oracle never panics at all, but if it does, the error would be recognizable.

Reply:
> The panic would be coming from the lines in the function body - so the call to the oracle module.

Then we need to revise where in x/oracle this could panic (as a result) of calling it and have it return an error instead.

Reply:
Removed the additional `recover()` logic, so the remaining change is just a check against negative/zero exchange rates. That should cover everything.